### PR TITLE
Copy for `ControlledQubitUnitary`

### LINKED ISSUE #1385: Copying `qml.ControlledQubitUnitary`
The `Operator` base class defines a `__copy__` method for operators.  For almost all `Operator`'s, this method works fine to duplicate a given instance.

But `qml.ControlledQubitUnitary` has a more complicated set of data, and copying it does not work fully.

```python
import pennylane as qml
from pennylane import numpy as np

import copy

op = qml.ControlledQubitUnitary(np.eye(2) * 1j, wires=[0], control_wires=[2])
print(op.matrix)

cop = copy.copy(op) # = op.__copy__()
print(cop.matrix)
```

Error
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-4a2b6ba0ebcf> in <module>
----> 1 cop.matrix

pennylane/pennylane/operation.py in matrix(self)
    666     @property
    667     def matrix(self):
--> 668         op_matrix = self._matrix(*self.parameters)
    669 
    670         if self.inverse:

pennylane/pennylane/ops/qubit.py in _matrix(self, *params)
   2203 
   2204     def _matrix(self, *params):
-> 2205         if self._CU is None:
   2206             self._CU = block_diag(np.eye(self._padding_left), self.U, np.eye(self._padding_right))
   2207 
```

To remedy this problem, I propose that this class defines its own `__copy__` method.

We should also test `op.__copy__()` for each operation type.  Note that just calling `__copy__()` is not sufficient to tell if there is a problem.  In this situation, you need to call methods like `op_copy.matrix` to see that the copy was insufficient.

Relevant locations in code:
* [``ControlledQubitUnitary``](https://github.com/PennyLaneAI/pennylane/blob/0dcbed53b470eb7f6f93b54077e36ca6287235cf/pennylane/ops/qubit.py#L2110)
* [``Operator.__copy__(self)``](https://github.com/PennyLaneAI/pennylane/blob/0dcbed53b470eb7f6f93b54077e36ca6287235cf/pennylane/operation.py#L255)

### PR DESCRIPTION
**Pull request by albi3ro**
Closes #1385 

This code now runs without error:

```python
import pennylane as qml
from pennylane import numpy as np

import copy

op = qml.ControlledQubitUnitary(np.eye(2) * 1j, wires=[0], control_wires=[2])
print(op.matrix)

cop = copy.copy(op) # = op.__copy__()
print(cop.matrix)
```

I have also added tests to prevent this sort of thing from happening again.

### CODE REVIEW

**Comment by albi3ro** [comment]
As this PR doesn't actually change the number of class attributes, we should ignore the pylint error for now.  But maybe that is something that  should be fixed in the future?

File: tests/ops/test_qubit_ops.py [hunk 1]
```diff
@@ -406,65 +407,86 @@ def circuit(basis_state):
...
+    qml.DoubleExcitationMinus(0.123, wires=[0, 1, 2, 3]),
+    qml.QubitSum(wires=[0, 1, 2]),
+]
+
 
 class TestOperations:
     """Tests for the operations"""
 
+    @pytest.mark.parametrize("op_cls, mat", NON_PARAMETRIZED_OPERATIONS)
+    def test_nonparametrized_op_copy(self, op_cls, mat, tol):
```
Comment (thisac) [inline tests/ops/test_qubit_ops.py:461]: Perhaps add tests that actual (deep) copies are being made? Or, if that's preferable, having `copy.deepcopy` make deep copies (testing that as well).

File: tests/ops/test_qubit_ops.py [hunk 2]
```diff
@@ -406,65 +407,86 @@ def circuit(basis_state):
     (qml.Toffoli, Toffoli),
 ]
 
+PARAMETRIZED_OPERATIONS = [
```
Comment (thisac) [inline tests/ops/test_qubit_ops.py:410]: 💯

File: pennylane/ops/qubit.py [hunk 3]
```diff
@@ -2201,6 +2201,17 @@ def __init__(self, *params, control_wires=None, wires=None, control_values=None,
...
+    def __copy__(self):
+        new_op = ControlledQubitUnitary(
+            self.U,
+            control_wires=self.control_wires,
+            wires=self._target_wires,
+            control_values=self.control_values,
+            do_queue=False,
+        )
+        new_op._inverse = self._inverse
+        return new_op
```
Comment (thisac) [inline pennylane/ops/qubit.py:2213]: Nice! 👍

**Review by thisac — APPROVED** [review]
Nice job @albi3ro! 💯 

I just have one thought/note on operator copies in general, which could be addressed here as well. It might be good to add (or at least keep in mind) that these tests do _not_ test if deep copies are performed, or if they even can be performed. A quick test actually shows that neither `copy.copy` nor `copy.deepcopy` actually copy the non-parametrized gate-matrices, instead only providing references. This seems to not be an issue for parametrized ops since their matrices aren't hard-coded, but it could be good to have this checked as well.

Even though the matrices shouldn't be changed by users, making this not a great deal really, it might cause weird issues and errors down the line.

**Comment by josh146** [comment]
@albi3ro it has suddenly struck me that there was another fix for this in an old WIP PR, here:

https://github.com/PennyLaneAI/pennylane/pull/1276/files#diff-ed5f9c0de17892f74e5119e2a032fe277eaa330ec4df0c249c7a75af7be6585eR260

This fix modified the parent `Operation.__copy__` to copy _all_ instance attributes (again, except for the operation parameters).

Would this fix work here?

**Comment by albi3ro** [comment]
@josh146 Seems like that would also work.  Should we just use that method instead?

**Comment by josh146** [comment]
Since it is a more general fix, I think that might be a good idea (to avoid similar bugs appearing in the future)

**Comment by albi3ro** [comment]
Instead of creating a operator specific copy, this PR now creates a more general operation copy.

**Review by thisac — APPROVED** [review]
Looks good. I'm still slightly confounded by why at least a deepcopy doesn't copy the matrix for non-parametrized operations, but I created an issue #1415 for it, in case it's something that we wish to look into later.

**Pull request merged**